### PR TITLE
Add panel fragments to read

### LINKED ISSUE #43: SAP2000_Toolkit: ReadPanel() to get additional properties as Fragment
#### Description:

Create Fragment for panels to capture:

- [ ] Auto Mesh
- [ ] Edge Constraint
- [ ] Material Overwrite
- [ ] Notional Size
- [ ] Offsets
- [ ] Transformation Matrix

### PR DESCRIPTION
**Pull request by JosefTaylor**
### NOTE: Depends on 

https://github.com/BHoM/Revit_Toolkit/pull/1130

   
### Issues addressed by this PR

Closes #43 

### Test files

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/SAP2000_Toolkit/%23241-Panel-Fragments?csf=1&web=1&e=qmCvdW

### CODE REVIEW

**Comment by JosefTaylor** [comment]
@BHoMBot check installer

**Comment by JosefTaylor** [comment]
@BHoMBot check copyright

**Comment by JosefTaylor** [comment]
@BHoMBot check copyright-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot fix copyright headers ref. 3557557496

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot fix project file ref. 4305558738

**Comment by JosefTaylor** [comment]
@BHoMBot check project-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check documentation-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot fix documentation punctuation ref. 4314388387

**Comment by JosefTaylor** [comment]
@BHoMBot check documentation-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check documentation-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check versioning

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check core

**Comment by JosefTaylor** [comment]
@BHoMBot check serialisation

**Comment by JosefTaylor** [comment]
@BHoMBot check installer

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

File: SAP2000_oM/Fragments/IPanelAutoMesh.cs
Comment (IsakNaslundBh) [inline SAP2000_oM/Fragments/IPanelAutoMesh.cs:37]: Not sure why all of these are on different lines?

Can we change all of them to be single line properties as the general convention. Especially when default values are assigned keeping them on multiple lines makes it really quite hard to read.

```suggestion
        bool LocalAxesOnEdge { get; set; }
```

**Review by IsakNaslundBh — DISMISSED** [review]
Think the oM properties needs adjusting in terms of formatting, apart from that, the code looks fine. Not able to test as I do not have SAP2000 though

**Comment by JosefTaylor** [comment]
@enarhi @mchaf would you do the honor of testing?

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
Fixed various reference issues lingering from the rebase, @IsakNaslundBh @enarhi @mchaf would you mind looking over?

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

File: SAP2000_Engine/Query/Panel.cs
Comment (IsakNaslundBh) [inline SAP2000_Engine/Query/Panel.cs:46]: This method will fail the nullcheck. Something like this should sort it I think. (if the panel is null, then the list of fragments will also be null from the `?` )

```suggestion
            List<IFragment> fragments = panel?.GetAllFragments(typeof(IPanelAutoMesh));
            
            if(fragments == null || fragments.Count == 0)
               retrun null;
            
            if (fragments.Count > 1) Compute.RecordWarning($"the panel {panel.Name} has more than one PanelOffset defined, which is not allowed. Only the first has been returned. Use GetAllFragments() to extract others.");
            return fragments.Select(x => x as IPanelAutoMesh).FirstOrDefault();
```

File: SAP2000_Engine/Query/Panel.cs
Comment (IsakNaslundBh) [inline SAP2000_Engine/Query/Panel.cs:64]: As above

```suggestion
            List<IFragment> fragments = panel?.GetAllFragments(typeof(IPanelOffset));
                        
            if(fragments == null || fragments.Count == 0)
               retrun null;
            
            if (fragments.Count > 1) Compute.RecordWarning($"the panel {panel.Name} has more than one PanelOffset defined, which is not allowed. Only the first has been returned. Use GetAllFragments() to extract others.");
            return fragments.Select(x => x as IPanelOffset).FirstOrDefault();
```

**Review by IsakNaslundBh — COMMENTED** [review]
Commetns on methods probably cause to the null handling failures

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance
@BHoMBot check null-handling

**Comment by IsakNaslundBh** [comment]
@JosefTaylor I had to rebase and force push this PR again after #253 was merged. as usual, please delete the branch locally before pulling down changes again! 🛩️

**Comment by JosefTaylor** [comment]
@BHoMBot check installer

**Comment by JosefTaylor** [comment]
@BHoMBot check versioning

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check core
@BHoMBot check serialization

**Comment by JosefTaylor** [comment]
@BHoMBot check serialisation

**Comment by enarhi** [comment]
@BHoMBot check required

**Review by enarhi — APPROVED** [review]
LGTM, tested using the provided test script and fragments were assigned properly, pushed to SAP, and pulled as expected.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance
@BHoMBot check ready-to-merge

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am authorising dispensation to be granted on check ref. 10538727097

**Comment by FraserGreenroyd** [comment]
Although there are new files, I have dispensated the check in favour of doing a smaller, newer, surgical PR tomorrow to update copyright headers again as this PR has been open a while and already rebased once so we don't want to keep it open too much longer.

**Pull request merged**
